source/docs/api-reference/wpilib-integration/unit-testing.rst

```diff
@@ -0,0 +1,137 @@
+
+In unit tests, users should utilize the ``StatusSignal.waitForUpdate()`` and ``BaseStatusSignal.waitForAll()`` APIs to wait for fresh data after sending a control request or modifying the simulated device state.
+
+.. important:: There may be a short delay between sending a control request and the simulated device applying the control. Unit tests should delay for ~20ms after sending a control request to account for this delay.
+
+Below is an example unit test that verifies the robot is enabled and verifies that the device responds to a control request.
+
+.. tab-set::
+
+   .. tab-item:: Java
+      :sync: Java
+
+      .. code-block:: java
+
+         public class TalonFXTest implements AutoCloseable {
+            static final double DELTA = 1e-3; // acceptable deviation range
+
+            TalonFX m_fx;
+            TalonFXSimState m_fxSim;
+
+            @Override
+            public void close() {
+               /* destroy our TalonFX object */
+               m_fx.close();
+            }
+
+            @BeforeEach
+            public void constructDevices() {
+               assert HAL.initialize(500, 0);
+
+               /* create the TalonFX */
+               m_fx = new TalonFX(0);
+               m_fxSim = m_fx.getSimState();
+
+               /* enable the robot */
+               DriverStationSim.setEnabled(true);
+               DriverStationSim.notifyNewData();
+
+               /* delay ~100ms so the devices can start up and enable */
+               Timer.delay(0.100);
+            }
+
+            @AfterEach
+            void shutdown() {
+               close();
+            }
+
+            @Test
+            public void robotIsEnabled() {
+               /* verify that the robot is enabled */
+               assertTrue(DriverStation.isEnabled());
+            }
+
+            @Test
+            public void motorDrives() {
+               /* set the voltage supplied by the battery */
+               m_fxSim.setSupplyVoltage(RobotController.getBatteryVoltage());
+
+               var dutyCycle = m_fx.getDutyCycle();
+
+               /* wait for a fresh duty cycle signal */
+               dutyCycle.waitForUpdate(0.100);
+               /* verify that the motor output is zero */
+               assertEquals(dutyCycle.getValue(), 0.0, DELTA);
+
+               /* request 100% output */
+               m_fx.setControl(new DutyCycleOut(1.0));
+               /* wait for the control to apply */
+               Timer.delay(0.020);
+
+               /* wait for a new duty cycle signal */
+               dutyCycle.waitForUpdate(0.100);
+               /* verify that the motor output is 1.0 */
+               assertEquals(dutyCycle.getValue(), 1.0, DELTA);
+            }
+         }
+
+   .. tab-item:: C++
+      :sync: C++
+
+      .. code-block:: cpp
+
+         class TalonFXTest : public testing::Test {
```

Comment:
Is this tested in a robot project? I never did any C++ unit tests before

Reply:
Yeah, I originally wrote it in a robot project to make sure it actually worked. I also just double-checked again that it worked because I couldn't remember if `SetUp()` worked or if I needed to use `SetUpTestCase()`.